### PR TITLE
Enable AlwaysPullImages without messing with the admission chain

### PR DESCRIPTION
**Pull request by deads2k**
This lets you enable the AlwaysPullImages admission plugin without changing the entire chain.

@csrwng ptal
@stenwt I think you were the one asking about this?

### CODE REVIEW

**Comment by csrwng** [comment]
@dak1n1 @abhgupta fyi

**Comment by deads2k** [comment]
[test]

**Comment by csrwng** [comment]
I don't think it's worth carrying an upstream change just so we can steal the name... can't we just name our wrapper plugin differently in origin?

File: pkg/cmd/server/api/types.go [hunk 1]
```diff
@@ -1193,3 +1193,14 @@ type ServiceServingCert struct {
...
 	Signer *CertInfo
 }
+
+// AdmissionPluginActivation can be used to enable to disable various admission plugins.
+// When this type is present as the `configuration` object in and `pluginConfig` and *if* the admission plugin supports it,
+// this will cause an off by default admission plugin to be enabled
+type AdmissionPluginActivation struct {
+	unversioned.TypeMeta
+
+	// TODO, we may want to make it possible to turn off an on by default admission plugin
```
Comment (csrwng) [inline pkg/cmd/server/api/types.go]: s/an/and/

File: pkg/cmd/server/api/v1/types.go [hunk 2]
```diff
@@ -1199,3 +1199,14 @@ type ServiceServingCert struct {
 	// If this value is nil, then certs are not signed automatically.
 	Signer *CertInfo `json:"signer"`
 }
+
+// AdmissionPluginActivation can be used to enable to disable various admission plugins.
+// When this type is present as the `configuration` object in and `pluginConfig` and *if* the admission plugin supports it,
```
Comment (csrwng) [inline pkg/cmd/server/api/v1/types.go]: s/in//

File: pkg/cmd/server/api/v1/types.go [hunk 3]
```diff
@@ -1199,3 +1199,14 @@ type ServiceServingCert struct {
...
+// this will cause an off by default admission plugin to be enabled
+type AdmissionPluginActivation struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// TODO, we may want to make it possible to turn off an on by default admission plugin
```
Comment (csrwng) [inline pkg/cmd/server/api/v1/types.go]: s/an/and/

File: pkg/image/admission/alwayspullimages/admission.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+	"k8s.io/kubernetes/pkg/admission"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/plugin/pkg/admission/alwayspullimages"
+
+	configlatest "github.com/openshift/origin/pkg/cmd/server/api/latest"
+)
+
+func init() {
+	admission.RegisterPlugin("AlwaysPullImages", func(client clientset.Interface, config io.Reader) (admission.Interface, error) {
```
Comment (csrwng) [inline pkg/image/admission/alwayspullimages/admission.go]: make a constant? (same string is used more than once)

File: test/integration/admissionconfig_test.go [hunk 5]
```diff
@@ -291,3 +291,80 @@ func TestOpenshiftAdmissionPluginEmbeddedConfig(t *testing.T) {
...
+	kubeConfigFile, err := testserver.StartConfiguredMaster(masterConfig)
+	if err != nil {
+		t.Fatalf("error starting server: %v", err)
+	}
+	kubeClient, err := testutil.GetClusterAdminKubeClient(kubeConfigFile)
+	if err != nil {
+		t.Fatalf("error getting client: %v", err)
+	}
+
+	if err := testserver.WaitForPodCreationServiceAccounts(kubeClient, "default"); err != nil {
```
Comment (csrwng) [inline test/integration/admissionconfig_test.go]: s/default/testutil.Namespace()/

File: test/integration/admissionconfig_test.go [hunk 6]
```diff
@@ -291,3 +291,80 @@ func TestOpenshiftAdmissionPluginEmbeddedConfig(t *testing.T) {
...
+	testPod.GenerateName = "test"
+	testPod.Spec.Containers = []kapi.Container{
+		{
+			Name:            "container",
+			Image:           "openshift/origin-pod:notlatest",
+			ImagePullPolicy: kapi.PullNever,
+		},
+	}
+
+	actualPod, err := kubeClient.Pods("default").Create(testPod)
```
Comment (csrwng) [inline test/integration/admissionconfig_test.go]: s/default/testutil.Namespace()/

File: test/integration/admissionconfig_test.go [hunk 7]
```diff
@@ -291,3 +291,80 @@ func TestOpenshiftAdmissionPluginEmbeddedConfig(t *testing.T) {
...
+	_, kubeConfigFile, err := testserver.StartTestMaster()
+	if err != nil {
+		t.Fatalf("error starting server: %v", err)
+	}
+	kubeClient, err := testutil.GetClusterAdminKubeClient(kubeConfigFile)
+	if err != nil {
+		t.Fatalf("error getting client: %v", err)
+	}
+
+	if err := testserver.WaitForPodCreationServiceAccounts(kubeClient, "default"); err != nil {
```
Comment (csrwng) [inline test/integration/admissionconfig_test.go]: testutil.Namespace()

File: test/integration/admissionconfig_test.go [hunk 8]
```diff
@@ -291,3 +291,80 @@ func TestOpenshiftAdmissionPluginEmbeddedConfig(t *testing.T) {
...
+	testPod.GenerateName = "test"
+	testPod.Spec.Containers = []kapi.Container{
+		{
+			Name:            "container",
+			Image:           "openshift/origin-pod:notlatest",
+			ImagePullPolicy: kapi.PullNever,
+		},
+	}
+
+	actualPod, err := kubeClient.Pods("default").Create(testPod)
```
Comment (csrwng) [inline test/integration/admissionconfig_test.go]: testutil.Namespace()

**Comment by deads2k** [comment]
> I don't think it's worth carrying an upstream change just so we can steal the name... can't we just name our wrapper plugin differently in origin?

We could, but I think this is likely to be the scout for the army.  I don't think kube will build a config like ours in the short term, which means that anything we want to conditionally enable ends up going down a path like this.  If we find a reasonable way to create a secondary "enabled" hook, we'll be using this to enable to disable significant numbers of plugins and we'll wish we didn't have an old one-off that we have to maintain forever for backward compatibility.

**Comment by deads2k** [comment]
@csrwng I think I'd be planning to do something clever with a secondary set of "default on, default off" lists that are applied inside of `InitPlugin` somehow and those are the plugins which are enabled this way.

That patch is harder to make, but would you like to see it now?

**Comment by csrwng** [comment]
@deads2k not necessarily. Regardless of how we plan to fix it, though, I guess I still don't see the need to introduce the additional carry.

**Comment by deads2k** [comment]
> @deads2k not necessarily. Regardless of how we plan to fix it, though, I guess I still don't see the need to introduce the additional carry.

If I introduce a new name, I have to maintain that name forever.  That name is effectively API (setting this part of the config turns on this feature) is a contract I'm not allowed to break.

See https://github.com/openshift/origin/pull/8945 for alternative implementation that also requires an upstream patch, but gives me the ability to control my admission chain without controlling the order.

**Pull request closed without merging**